enhancements/installer/cluster-api-based-installations.md

```diff
@@ -0,0 +1,485 @@
+kind: AWSCluster
+metadata:
+  creationTimestamp: null
+  name: capi-example-kbdpg
+  namespace: openshift-cluster-api-guests
+spec:
+  bastion:
+    enabled: false
+  controlPlaneEndpoint:
+    host: ""
+    port: 0
+  controlPlaneLoadBalancer:
+    additionalListeners:
+    - port: 22623
+      protocol: TCP
+    crossZoneLoadBalancing: false
+    loadBalancerType: nlb
+    name: capi-example-kbdpg-ext
+    scheme: internet-facing
+  network:
+    additionalControlPlaneIngressRules:
+    - description: MCS traffic from cluster network
+      fromPort: 22623
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - node
+      - controlplane
+      toPort: 22623
+    - description: controller-manager
+      fromPort: 10257
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - controlplane
+      - node
+      toPort: 10257
+    - description: kube-scheduler
+      fromPort: 10259
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - controlplane
+      - node
+      toPort: 10259
+    cni:
+      cniIngressRules:
+      - description: ICMP
+        fromPort: -1
+        protocol: icmp
+        toPort: -1
+       # several rules omitted for brevity
+      - description: Service node ports (UDP)
+        fromPort: 30000
+        protocol: udp
+        toPort: 32767
+    vpc:
+      availabilityZoneSelection: Ordered
+      availabilityZoneUsageLimit: 3
+      cidrBlock: 10.0.0.0/16
+  region: us-east-2
+  s3Bucket:
+    name: openshift-bootstrap-data-capi-example-kbdpg
+    presignedURLDuration: 1h0m0s
+status:
+  networkStatus:
+    apiServerElb:
+      attributes: {}
+  ready: false
+```
+
+This manifest describes the network, load balancers, security group rules, etc. that will be used or provisioned for
+the target cluster. Control plane machines are provisioned by [AWSMachines][AWSMachines] & [CAPI Machines][CAPIMachines]
+manifests.
+
+##### Additional Infrastructure
+
+The Installer will generate any additional infrastructure that is needed but not handled by the CAPI provider, either
+because it is out of scope (e.g. IAM, DNS) or not adopted upstream (e.g. split-horizon load balancers). The Installer
+will provide hooks into the provisioning lifecycle that can be used to provision resources using direct SDK calls
+or other tooling.
+
+The AWS proof-of-concept implementation utilizes hooks defined in this interface:
+
+```go
+type CAPIInfraHelper interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun.
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(in PreProvisionInput) error
+
+	// ControlPlaneAvailable is called once cluster.Spec.ControlPlaneEndpoint.IsValid()
+	// returns true, typically after load balancers have been provisioned. It can be used
+	// to create DNS records.
+	ControlPlaneAvailable(in ControlPlaneAvailableInput) error
+}
+
+type PreProvisionInput struct{ clusterID string }
```

Comment:
In OpenStack, we would need more than just the `clusterID` to properly prepare the infrastructure. We'd also need to manipulate the CAPI manifests (most notably: there is a hard limit on the allowed size for the Ignition file, which means that we must upload it in the infrastructure and replace it with a shim in the bootstrap manifest).

Maybe the `PreProvisionInput` struct could contain the collection of the Installer assets available at this point?

Reply:
Are you expecting to be able to inject that token before or after the `manifests` phase of the install process?

Reply:
In [this PoC](https://github.com/openshift/installer/pull/7856), I expect the generated CAPI bootstrap Machine to contain the full Ignition file (the one that is too big to be uploaded as Nova userdata). Then in the cluster-create manifest, I upload it to OpenStack and replace it in the Machine with a shim containing a link to the uploaded one, complete with an authorization header containing the ephemeral token.

As a side-effect, this process enables the user to see and modify the full Ignition file before we use it.

The PoC proposes a simplified version of the pre-capi hook proposed in this RFE paragraph, with a call to `PreCAPI` that is identical to the existing call to `PreTerraform`. [Its function signature contains a pointer to the manifests](https://github.com/openshift/installer/pull/7856/files#diff-3efede313acff180e1dd57d61faa806e09bce90c4779793fb1bf155160c705a9R175).